src/main/java/io/vertx/core/net/impl/SSLHelper.java

```diff
@@ -409,36 +429,65 @@ private SslHandler createHandler(SSLEngine engine, boolean client) {
       sslParameters.setEndpointIdentificationAlgorithm(endpointIdentificationAlgorithm);
       engine.setSSLParameters(sslParameters);
     }
-    return new SslHandler(engine);
+    if (serverName != null) {
+      SSLParameters sslParameters = engine.getSSLParameters();
+      sslParameters.setServerNames(Collections.singletonList(new SNIHostName(serverName)));
+      engine.setSSLParameters(sslParameters);
+    }
   }
 
   public SslContext getContext(VertxInternal vertx) {
-    if (sslContext == null) {
-      sslContext = createContext(vertx);
-      if (sslContext instanceof OpenSslServerContext){
-        SSLSessionContext sslSessionContext = sslContext.sessionContext();
-        if (sslSessionContext instanceof OpenSslServerSessionContext){
-          ((OpenSslServerSessionContext)sslSessionContext).setSessionCacheEnabled(openSslSessionCacheEnabled);
-        }
+    return getContext(vertx, null);
+  }
+
+  public SslContext getContext(VertxInternal vertx, String serverName) {
+    if (serverName == null) {
+      if (sslContext == null) {
+        sslContext = createContext(vertx, null);
+      }
+      return sslContext;
+    } else {
+      X509KeyManager mgr;
+      try {
+        mgr = keyCertOptions.keyManagerMapper(vertx).apply(serverName);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
       }
+      if (mgr == null) {
+        return sslContext;
```

Comment:
yes this is deliberate, SSLHelper#getContext(null) is always called by the `validate` method when the server is starting. It could probably be improved to be more readable I think with a small refactoring.